### PR TITLE
Feature/endpoints

### PR DESCRIPTION
**Pull request by RWOverdijk**
Woa, big one.
- Added tests
- Added endpoint support (configure your own client)
- Added support to patch (configure) multiple endpoints to send along the Authorization header
- Added typescript d.ts generation
- Stuff. :dancers:

### CODE REVIEW

File: src/app.fetch-httpClient.config.js [hunk 1]
```diff
@@ -3,41 +3,84 @@ import {Authentication} from './authentication';
...
-
-              request.headers.append(config.authHeader, token);
-            }
-
-            return request;
-          }
-        });
+
+    return {
+      request(request) {
```
Comment (doktordirk) [inline src/app.fetch-httpClient.config.js]: request: function(request)?

File: src/baseConfig.js [hunk 2]
```diff
@@ -11,6 +11,8 @@ export class BaseConfig {
 
   constructor() {
     this._current = {
+      endpoint: null, // Endpoint to use for auth requests (local only). Null uses HttpClient.
```
Comment (doktordirk) [inline src/baseConfig.js:14]: i got a bit confused when looking up _current.client. can it be added here to make it easier to find?
like client: null, // the currently used client. gets set during configuration (see index.js)
or alike

Comment (RWOverdijk) [inline src/baseConfig.js:14]: You shouldn't touch client. I use that internally. You _could_ use it, but you're not supposed to.

File: src/app.fetch-httpClient.config.js (hunk 1 above)
Comment (RWOverdijk) [inline src/app.fetch-httpClient.config.js:38]: It's a getter. Try it: 

``` js
var foo = {
  bar (name) {
    console.log('Hello ' + name);
  }
};

foo.bar('Dirk');
```

File: src/baseConfig.js (hunk 2 above)
Comment (doktordirk) [inline src/baseConfig.js:14]: i assumed so. unfortunatelly i removed the line in the comment where i explicitly said said. it's just when i browsed the code i saw config.client and wondered what it was, so i checked here and didn't see it either. only later i spotted it in index.js. hence, the suggestion to add a comment here. like (old version):
client: null, // for internal use. the currently used rest client. gets set during configuration (see index.js)

Comment (RWOverdijk) [inline src/baseConfig.js:14]: I can live with that. I'll add that, good suggestion.

Comment (RWOverdijk) [inline src/baseConfig.js:14]: I've [added it](https://github.com/RWOverdijk/aurelia-auth/commit/58e9fa6d4753d28a656fc3e093d3e3205bb05159). This makes it clearer for weary travelers :D :+1:

File: src/index.js [hunk 3]
```diff
@@ -1,15 +1,41 @@
...
+    baseConfig.current.configureEndpoints.forEach(endpointToPatch => {
+      fetchConfig.configure(endpointToPatch);
+    });
+  }
+
+  let client = clientConfig.getEndpoint(baseConfig.current.endpoint);
+
+  if (!(client instanceof Rest)) {
+    client = new Rest(aurelia.container.get(HttpClient));
+  }
```
Comment (doktordirk) [inline src/index.js]: maybe 
in baseconfig:
configureEndpoints: false, // false to skip, true to configure default http-fetch-client resp. default rest client. string or array of strings to configure selected endpoints

if (endpointToPatch) {
  let configureEndpoints = baseConfig.current.configureEndpoints === true ? client : baseConfig.current.configureEndpoints;
  fetchConfig.configure(configureEndpoints);
}

File: src/index.js [hunk 4]
```diff
@@ -1,15 +1,41 @@
...
     config(baseConfig);
   } else if (typeof config === 'object') {
     baseConfig.configure(config);
   }
+
+  if (Array.isArray(baseConfig.current.configureEndpoints)) {
+    baseConfig.current.configureEndpoints.forEach(endpointToPatch => {
+      fetchConfig.configure(endpointToPatch);
+    });
+  }
```
Comment (doktordirk) [inline src/index.js]: remove

**Pull request merged**